user/src/com/google/gwt/user/server/rpc/RemoteServiceServlet.java

```diff
@@ -366,23 +373,61 @@ public final void processPost(HttpServletRequest request,
     // Let subclasses see the serialized request.
     //
     onBeforeRequestDeserialized(requestPayload);
-
-    // Invoke the core dispatching logic, which returns the serialized
-    // result.
-    //
-    String responsePayload = processCall(requestPayload);
-
-    // Let subclasses see the serialized response.
+    
+    // Create Writer on response, considering GZIP compression if requested
     //
-    onAfterResponseSerialized(responsePayload);
+    final Writer responseWriter = createWriterForResponse(request, response);
+    final boolean onAfterResponseSerializedOverridden = isOnAfterResponseSerializedOverridden();
+    final StringWriter writerForResponseCopy;
+    if (onAfterResponseSerializedOverridden) {
+      writerForResponseCopy = new StringWriter();
+    } else {
+      writerForResponseCopy = null;
+    }
+    final TeeWriter<StringWriter> writer = new TeeWriter<>(responseWriter, writerForResponseCopy);
+    RPC.setResponseWriter(writer);
+    try {
+      // Invoke the core dispatching logic, which returns the serialized
+      // result.
+      //
+      String responsePayload = processCall(requestPayload);
+      
+      if (onAfterResponseSerializedOverridden) {
+        // Let subclasses see the serialized response.
+        // For the old protocol version it was already assembled
+        // and returned by processCall(requestPayload). For protocol
+        // versions starting with AbstractSerializationStream.SERIALIZATION_STREAM_FORWARD_STREAMING_VERSION
+        // null is returned by processCall(requestPayload), and the
+        // payload is obtained from the StringWriter
+        onAfterResponseSerialized(responsePayload != null ? responsePayload : writerForResponseCopy.toString());
+      }
+    } finally {
+      RPC.unsetResponseWriter();
+    }
+  }
 
-    // Write the response.
-    //
-    writeResponse(request, response, responsePayload);
+  private boolean isOnAfterResponseSerializedOverridden() {
+    try {
+      Class<?> c = getClass();
+      boolean foundOverride = false;
+      while (!foundOverride && c != RemoteServiceServlet.class) {
```

Comment:
devil's advocate: what if this method is implemented by an interface with a default method?

(IDEA believes that the `m != null` check is always true, but I'm not sure it is right...)

Reply:
I think then the RemoteServiceServlet implementation would "hide" the default method illegally. Eclipse reports

"The inherited method RemoteServiceServlet.onAfterResponseSerialized(String) cannot hide the public abstract method in OnAfterResponseSerialized"

when I try this, and this makes sense to me. So I cannot see a way how a (public) default interface method could provide an implementation for onAfterResponseSerialized successfully.